packages/ymax-planner/src/cosmos-rpc.ts

```diff
@@ -0,0 +1,158 @@
+      const sub = this.#subscriptions.get(response.id);
+      if (sub) {
+        sub.notified(response);
+      }
+      this.receive(response);
+    });
+
+    ws.addEventListener('open', () => {
+      this.#openedPK.resolve();
+    });
+  }
+
+  receive(response: JSONRPCResponse) {
+    // console.log('Received RPC response:', response);
+    return super.receive(response);
+  }
+
+  close() {
+    this.rejectAllPendingRequests(`RPC client closed`);
+    this.#ws.close();
+    this.#subscriptions.clear();
+  }
+
+  opened() {
+    return this.#openedPK.promise;
+  }
+
+  closed() {
+    return this.#closedPK.promise;
+  }
+
+  subscribe(query: string) {
+    return this.subscribeAll([query]);
+  }
+
+  async *subscribeAll(queries: string[]) {
+    type Cell = { head: JSONRPCResponse; tail: Promise<Cell> };
+    let lastPK = Promise.withResolvers<Cell>();
+    let firstCell: Promise<Cell> = lastPK.promise;
+
+    const newQueriesSet = new Set(queries);
+    if (newQueriesSet.size === 0) {
+      throw new Error(`No new queries to subscribe to: ${queries.join(', ')}`);
+    }
+
+    const currentQueriesCount = this.#subscriptions.size;
+    const totalQueries = newQueriesSet.size + currentQueriesCount;
+    if (totalQueries > MAX_SUBSCRIPTIONS) {
+      throw new Error(
+        `Too many subscriptions: new ${newQueriesSet.size} + existing ${currentQueriesCount} exceeds RPC client limit of ${MAX_SUBSCRIPTIONS}`,
+      );
+    }
+
+    const qs = [...newQueriesSet.keys()].map(query => {
+      const subP = this.request('subscribe', { query });
+      const subId = this.#lastSentId;
+      // console.log(`Subscribing to query "${query}" with ID ${subId}`);
+      const unsubscribe = () => {
+        this.#subscriptions.delete(subId);
+        void this.request('unsubscribe', { query });
+      };
+      let isFirstResponse = true;
+      this.#subscriptions.set(subId, {
+        query,
+        notified: (response: JSONRPCResponse) => {
+          if (isFirstResponse) {
+            // Ignore the initial empty data response.
+            isFirstResponse = false;
+            return;
+          }
+          // console.log('notified for query:', query, response);
+          const thisPK = lastPK;
+          lastPK = Promise.withResolvers<Cell>();
+          thisPK.resolve({ head: response, tail: lastPK.promise });
+        },
+        unsubscribe,
+      });
+      return { subP, unsubscribe };
+    });
+
+    // console.log(`Awaiting subscription responses for queries:`, qs);
+    await Promise.all(qs.map(({ subP }) => subP));
+
+    // console.log('wait forever?');
+    try {
+      while (true) {
+        // console.log('wait for next event');
+        const entry = await firstCell;
+        // console.log('got next event', entry);
+        firstCell = entry.tail;
+        if (entry.head.error) {
+          yield Promise.reject(Error(entry.head.error.message));
+          continue;
+        }
+        yield entry.head.result;
```

Comment:
`nextCell` seems like a more accurate name for this ever-advancing variable.
```suggestion
        const { head, tail } = await nextCell;
        nextCell = entry.tail;
        if (head.error) {
          // Propagate errors non-fatally.
          yield Promise.reject(Error(head.error.message));
          continue;
        }
        yield head.result;
```